raft/server.cc

```diff
@@ -1292,7 +1305,21 @@ future<> server_impl::set_configuration(config_member_set c_new, seastar::abort_
     // would be the one corresponding to our joint configuration,
     // no matter if the leader changed in the meantime.
 
-    auto f = _non_joint_conf_commit_promise.emplace().get_future();
+    auto guard = defer([this] {
+        // If everything goes right, `io_fiber` will reset the field,
+        // but in case of failure - ensure it's reset here.
+        _non_joint_conf_commit_promise.reset();
+    });
+    auto f = _non_joint_conf_commit_promise.emplace().promise.get_future();
+    if (as) {
+        _non_joint_conf_commit_promise->abort = as->subscribe([this] () noexcept {
+            // The subscription is destroyed when the field is reset, so if we're here, the field must be engaged.
+            // Also, whoever resolves the promise must reset the field. Thus, if we're here, the promise is not resolved.
+            assert(_non_joint_conf_commit_promise);
```

Comment:
As the comment says:
```
The subscription is destroyed when the field is reset, so if we're here, the field must be engaged.
```

Reply:
The comment is not the code. If I could avoid bugs by just writing comments I would not have any. The question is do we run all subscritions immediately and they cannot preempt when `as` is signaled? Because if not we can have a subscription that is scheduled to run and then we get an operation completion which resets `_non_joint_conf_commit_promise` before subscription runs.

Reply:
> 
> (also I briefly looked through abort source code and there are no yields between iterating over subscriptions and calling the callbacks so we should be good)

Can callbacks yield? Is this something `abort_source` as an API guarantied. We can easily handle the possible race here instead of aborting and not worry about it.

Reply:
> The comment is not the code. If I could avoid bugs by just writing comments I would not have any.

In fact you can avoid bugs by writing comments (specifically, by writing correctness proofs in your comments). But people rarely do that and they write many bugs. Here, I wrote a short proof that shows correctness of this assertion. I agree there is an implicit assumption in the proof (that the subscription still exists when the callback is called), but see below.

> The question is do we run all subscritions immediately and they cannot preempt when as is signaled? Because if not we can have a subscription that is scheduled to run and then we get an operation completion which resets _non_joint_conf_commit_promise before subscription runs.

There are three other calls to `subscribe` in raft/server.cc, two written by you.
All those uses of abort source assume that you at the moment subscription is called, it still exists. They use the same "trick" I use here.
So why are you doing the bikeshedding now, not when you wrote analogous code?

I'm guessing there are many other uses of abort_source which assume this.

abort_source should probably explicitly state this guarantee in its API.

Reply:
> > The comment is not the code. If I could avoid bugs by just writing comments I would not have any.
> 
> In fact you can avoid bugs by writing comments (specifically, by writing correctness proofs in your comments). But people rarely do that and they write many bugs. Here, I wrote a short proof that shows correctness of this assertion. I agree there is an implicit assumption in the proof (that the subscription still exists when the callback is called), but see below.
> 
There is an implicit assumption in your proof that a fiber running subscription cannot race with any other fiber. It may well be true.

> > The question is do we run all subscritions immediately and they cannot preempt when as is signaled? Because if not we can have a subscription that is scheduled to run and then we get an operation completion which resets _non_joint_conf_commit_promise before subscription runs.
> 
> There are three other calls to `subscribe` in raft/server.cc, two written by you. All those uses of abort source assume that you at the moment subscription is called, it still exists. They use the same "trick" I use here. So why are you doing the bikeshedding now, not when you wrote analogous code?
> 
You are saying as if I never do bugs :). I either did not think about it or looked into the subscription code and decided like you that the race is not possible. If I did not look and did not write the code defensively it is my fault.  So I looked now. Request aborting looks like that:

```
    void do_request_abort(std::optional<std::exception_ptr> ex) noexcept {
        assert(_subscriptions);
        _ex = ex.value_or(get_default_exception());
        auto subs = std::exchange(_subscriptions, std::nullopt);
        while (!subs->empty()) {
            subscription& s = subs->front();
            s.unlink();
            s.on_abort(ex);
        }
    }
```

The function returns `void`, but we cannot assume it cannot sleep unfortunately. If a notification is done from a seastar thread and a subscription before ours assumes it runs in a thread and calls `yield()` the whole loop  can yield. But is it a problem. `_subscriptions` list is ` bi::list<subscription, bi::constant_time_size<false>>` and subscription itself is `bi::list_base_hook<bi::link_mode<bi::auto_unlink>>`, so even if a subscription before ours yields if our subscription is destroyed meanwhile it should remove itself from  copied `subs` list and will not run.

> I'm guessing there are many other uses of abort_source which assume this.
> 
> abort_source should probably explicitly state this guarantee in its API.

It should.

Reply:
> There is an implicit assumption in your proof that a fiber running subscription cannot race with any other fiber.

It's enough to assume that either:
1. callback is never called, or
2. callback finishes before subscription is destroyed.

And the abort_source implementation must guarantee that.
As you've shown, it guarantees this for callbacks that don't yield.

Then the proof goes as follows:
- we're currently inside the callback
- from 2 above, subscription wasn't destroyed yet
- since sub wasn't destroyed, the optional wasn't reset (the sub lives inside the optional).

Created issue: https://github.com/scylladb/seastar/issues/1186